Html printing cleanup

vadz opened the pull request:
I was trying to understand how pagination in HTML printing worked in order to use it in my own code and noticed a few things that seemed simple to fix, so I did it (there are quite a few more remaining though...).

There should be no changes in behaviour.

@vslavik I'm not going to request a review from you because it's not that important but any comments would still be welcome, of course. TIA!

vadz committed:
Make wxHtmlPrintout::m_Renderer{,Hdr} objects instead of pointers

There is no need to allocate these objects on the heap and delete them
later when we can just make them members of wxHtmlPrintout object
itself.

No real changes.

vadz committed:
Remove unnecessary wxHtmlPrintout members initialization

Don't initialize strings to their default (empty) value, this is just
useless.

vadz committed:
Use "for" loop in wxHtmlContainerCell::AdjustPagebreak()

Use a simpler form for iterating over all children.

No real changes.

vadz committed:
Pass known page breaks as const ref to wxHtmlDCRenderer::Render()

This method doesn't modify the known page breaks, it only uses them to
find the next one, so make it more clear by using const reference in the
function signature.

vadz committed:
Use wxDCClipper in wxHtmlDCRenderer code

Use wxDCClipper instead of manually calling DestroyClippingRegion().
This is simpler and safer and also potentially more correct as
wxDCClipper restores the previous clipping region, if any, unlike the
old code -- even though this probably doesn't matter here as the
printing DC seems unlikely to be clipped.

vadz committed:
Add wxPrintout::SetUp() to reuse the same code in all ports

MSW, GTK, OSX and Postscript implementations all did almost exactly the
same thing to initialize wxPrintout, so extract this common code into a
new wxPrintout method and just call it instead.

There should be no changes in behaviour.

vadz committed:
Add a simple unit test for wxHtmlPrintout pagination logic

Verify that the number of pages to be printed is as expected.

vadz committed:
Slightly simplify wxHtmlDCRenderer::Render()

Use "totalHeight" variable to make it more clear that
m_Cells->GetHeight() and GetTotalHeight() are the same value.

vadz committed:
Split and simplify wxHtmlDCRenderer::Render()

This function was difficult to understand as it did two quite different
things depending on the value of its "dont_render" argument and using it
also made CountPages() logic more complicated than necessary.

Simplify the code by splitting Render() into FindNextPageBreak(), which
is used by CountPages(), and Render() itself, which doesn't need to deal
with pagination at all as it's either already provided the page start
and end positions or can just assume that everything fits on a single
page (the latter is the case for the headers and footers renderer).

This is a notionally backwards-incompatible change, but no code using
wxHtmlDCRenderer could be found in the wild and the new API is so much
simpler that it seems to be worth switching to it.

vadz committed:
Add a test of explicit page breaks in wxHtmlPrintout too

Check that using the style forcing a page break does work.

vadz committed:
Don't call AdjustPagebreak() in a loop, this seems useless

There doesn't seem to be any reason to call this function more than once
on the same cell as the existing implementations are idempotent and it's
difficult to see why this should ever not be the case.

vadz committed:
Simplify wxHtmlPageBreakCell::AdjustPagebreak()

There is no need to look in the known page breaks array for a page break
at the position of this cell, all we care about is whether this cell is
on the current page or not: we must insert a page break if, and only if,
it is.

No real changes in behaviour, but the code is now much simpler to
understand and this change paves way for removing "known_pagebreaks"
entirely, as it was only added for the use in this function (see
f2034f1b6e04a39f6d882cf6cdd53c11bf39f5d7 from 15 years ago), which
doesn't even actually need it.

It should also make pagination code somewhat faster.

vadz committed:
Add a sanity check for AdjustPagebreak() implementation

Verify that this function never adjusts page break so far back that it
comes before the previous one (or even at the same position).

This avoids infinite loops in CountPages() even if a custom cell class
implements its overridden AdjustPagebreak() incorrectly.

vadz committed:
Remove "known_pagebreaks" from wxHtmlCell::AdjustPagebreak()

This parameter is not actually needed for any reasonable page breaking
algorithm, as it shouldn't care about any previous page breaks except
for the last one, which is already known as "*pagebreak - pageHeight" in
this function.

Removing it will allow to stop using wxArrayInt in the code using this
method and switch to an std::vector<int> instead.

This is not a backwards compatible change, but it seems that there is
very little code actually overriding this function (none could be found
in any open source repositories) and updating it should be as simple as
removing the now unneeded argument.

vadz committed:
Use wxVector<int> instead of wxArrayInt in wxHtmlDCRenderer

No real changes, just prefer standard-like template class to the legacy
wx one.

vadz committed:
Replace wxList used in wxHtmlPrintout with wxVector<> too

Prefer the use of the standard-like template class to macro-based list.

It also makes more sense to use a vector rather than a linked list here,
as the elements are never removed from or inserted into m_Filters, so
there is no reason to prefer the list structure.

vadz committed:
Make wxHtmlDCRenderer::FindNextPageBreak() const

It doesn't modify the renderer.

vadz committed:
Document wxHtmlDCRenderer::FindNextPageBreak()

This is a useful (even if mostly trivial) function, so document it and
mention it in Render() documentation too.

vadz committed:
Remove redundant wxHtmlDCRenderer::m_NumPages

The actual number of pages is given by the number of elements in
m_PageBreaks vector after CountPages() is shown and is unknown until
then, so m_NumPages is completely unnecessary and can be just removed.

vadz commented:
Vaclav, sorry but I'm going to change my mind and ask you to review this, after all, as after looking at this code again and again, I realized that it was much more complicated and less clear than it could be and so decided to simplify it. The most important changes are the refactoring of `wxHtmlDCRenderer::Render()` into `FindNextPageBreak()` and `Render()` itself in e01892c669 and then getting rid of the array/vector of known page breaks in b9b6ccb804 as, IMO, it's completely unnecessary. Another thing is stopping calling `AdjustPageBreak()` in the loop and removing insistent mentions of having to do it from the comments/documentation in 63add2cd19 because, to the best of my knowledge, this can never do any good (well, anything better than calling it just once). Please let me know if I'm missing anything here, TIA!

vslavik reviewed (commented):
Looks OK to me, but don't put too much weight on that: I don't really remember any of this code, it's a depressing trip down the memory lane more than anything else.

vslavik commented:
> and then getting rid of the array/vector of known page breaks in b9b6ccb as, IMO, it's completely unnecessary. 

I'm not 100% sure of that, but I trust you tested the change. I *think* the motivation for that was the degenerated case of multiple subsequent such breaks and trying to prevent insertion of empty pages between them. I don't know what the *correct* behavior is supposed to be, though: the spec says the `<div>` that uses `page-break-before` cannot be empty, but how should something like
```html
<div style="page-break-before:always">
  <div style="page-break-before:always">
    <div style="page-break-before:always">
      on snap
    </div>
  </div>
</div>
```
behave is unclear to me.

> Another thing is stopping calling AdjustPageBreak() in the loop and removing insistent mentions of having to do it from the comments/documentation in 63add2c because, to the best of my knowledge, this can never do any good

Again just a hypothesis, but I think the motivation was that *if* `AdjustPagebreak()` could move the pagebreak in either direction, then multiple passes would be necessary. But as long as it's moved only backwards (as is the case with the existing implementations), the 2018 me doesn't see why single pass wouldn't suffice either.

vadz commented on src/html/htmprint.cpp:
```diff
@@ -199,7 +199,6 @@ wxVector<wxHtmlFilter*> wxHtmlPrintout::m_Filters;
 
 wxHtmlPrintout::wxHtmlPrintout(const wxString& title) : wxPrintout(title)
 {
-    m_NumPages = INT_MAX;
```
AFAICS it wasn't, which is one of the reasons why this code was rather confusing. Note that I still preserved returning `INT_MAX` in `GetPageInfo()`, which seems to be the only thing it was used for, even though I'm not really sure if `GetPageInfo()` can be called before `OnPreparePrinting()` (wxGTK code is the only place where it's not totally clear where this happens).

vadz commented:
I've tested (with the new unit test and in the `html/printing` sample) with the things I could think of, but I didn't think about your example above, for example. Testing it now, I think it does behave correctly: as written, there are no page breaks at all because there is already an implicit page break at position 0. If I add something before it, there is similarly just a single page break because nested divs are at position 0 of their parent and so don't create page breaks, i.e. this test passes:
```
    pr.SetHtmlText
       (
            "Something"
            "<div style=\"page-break-before:always\">"
                "<div style=\"page-break-before:always\">"
                    "<div style=\"page-break-before:always\">"
                        "whatever"
                    "</div>"
                "</div>"
            "</div>"
       );
    CHECK( CountPages(pr) == 2 );
```

I think this is the expected behaviour, isn't it?

As for moving page breaks forward: I hadn't found any version (and I looked at a few different iterations of this code all the way back to the time it was written) where it worked like this. I.e. it only ever moved backwards/higher and this also makes sense to me.

Thanks for reviewing and I'll merge this soon if there are no other comments.

BTW, the next thing I plan to do is to implement [`page-break-inside: avoid`](https://developer.mozilla.org/en-US/docs/Web/CSS/page-break-inside) support. AFAICS it should be simple using `SetCanLiveOnPagebreak()`, but I haven't tested it yet.

vadz committed:
Add another page-break-before test to pagination unit test

Check that the nested elements with this style are handled as expected.

vadz committed:
Fix HTML pagination test for high DPI values

The standard margins, expressed in millimeters, could result in the
usable page space being much smaller than 1000px used for the DC size
when using higher DPIs, which means that the test checking that a 2400px
image took only 3 pages could fail, as it could require 4 of them in
this case.

Fix this by getting rid of the margins, as this should ensure that the
page height is exactly 1000px now, independently of the actual DPI.

vslavik commented:
> I think this is the expected behaviour, isn't it?

I'd say so, yes.

> Thanks for reviewing and I'll merge this soon if there are no other comments.

There aren't, and the above ones were meant just as explanatory for things that (rightly) made no sense to you.

Thanks for cleaning up this part!
